lib/Sema/SemaBounds.cpp

```diff
@@ -628,93 +726,209 @@ namespace {
+          // if both `e1` and `e2` have bounds, then they must be equal.
+          // `e1` is an LValue, its bounds are the lvalue target bounds.
+          // `e2` is an RValue
+          //
+          // @ can stand for: +, -, *, /, %, &, |, ^, >>, <<
           if (LHS->getType()->isIntegerType() &&
               RHS->getType()->isIntegerType() &&
               (BinaryOperator::isAdditiveOp(Op) ||
                BinaryOperator::isMultiplicativeOp(Op) ||
-               BinaryOperator::isBitwiseOp(Op))) {
+               BinaryOperator::isBitwiseOp(Op) ||
+               BinaryOperator::isShiftOp(Op))) {
             BoundsExpr *LHSBounds = IsCompoundAssignment ?
               LValueTargetBounds(LHS) : RValueBounds(LHS);
             BoundsExpr *RHSBounds = RValueBounds(RHS);
             if (LHSBounds->isNone() && !RHSBounds->isNone())
               return RHSBounds;
             if (!LHSBounds->isNone() && RHSBounds->isNone())
               return LHSBounds;
-            // TODO: check if both LHS and RHS have bounds and
-            // indicate cause of failure to have bounds.
+            if (!LHSBounds->isNone() && !RHSBounds->isNone()) {
+              // TODO: Check if LHSBounds and RHSBounds are equal.
+              // if so, return one of them. If not, return bounds(none)
+              return CreateBoundsUnknown();
+            }
+            if (LHSBounds->isNone() && RHSBounds->isNone())
+              return CreateBoundsEmpty();
+          }
+
+          // Comparisons and Logical Ops
+          //
+          // `e1 @ e2` have empty bounds if @ is:
+          // ==, !=, <=, <, >=, >, &&, ||
+          if (BinaryOperator::isComparisonOp(Op) ||
+              BinaryOperator::isLogicalOp(Op)) {
+            return CreateBoundsEmpty();
           }
-          return CreateBoundsNone();
+
+          // All Other Binary Operators we don't know how to deal with
+          return CreateBoundsEmpty();
+        }
+        case Expr::CallExprClass: {
+          const CallExpr *CE = dyn_cast<CallExpr>(E);
+          if (!CE) {
+            llvm_unreachable("unexpected cast failure");
+            return CreateBoundsInferenceError();
+          }
+
+          // Get function prototype.
+          // The Callee is a function *pointer*, almost always.
+          const FunctionProtoType* CalleeTy = dyn_cast<FunctionProtoType>(CE->getCallee()->getType()->getPointeeType());
+          if (!CalleeTy)
+            // K&R functions have no prototype, and we cannot perform inference on them,
+            // so we return bounds(none) for their results.
+            return CreateBoundsEmpty();
+
+          BoundsExpr *FunBounds = const_cast<BoundsExpr *>(CalleeTy->getReturnBounds());
+          if (!FunBounds)
+            // This function has no return bounds
+            return CreateBoundsEmpty();
+
+          if (FunBounds->isInteropTypeAnnotation())
+            return FunBounds;
```

Comment:
From the other `->isInteropTypeAnnotation()` if-statement, I attempted to make this some equivalent to `bounds(none)` for conservativeness, but it was failing existing tests, so I went back to returning the interop type. 

I've now changed everywhere we come across an interop type annotation to return `CreateBoundsAllowedButNotComputed()`, including in `BoundsExpr *ExpandToRange(Expr *Base, BoundsExpr *B)`